runtime/gc_base/ObjectAccessBarrier.hpp

```diff
@@ -119,7 +119,13 @@ class MM_ObjectAccessBarrier : public MM_BaseVirtual
 		U_32 slotsPerArrayletLeaf = (U_32)(J9VMTHREAD_JAVAVM(vmThread)->arrayletLeafSize / elementSize);
 		U_32 arrayletIndex = (U_32)index / slotsPerArrayletLeaf;
 		U_32 arrayletOffset = (U_32)index % slotsPerArrayletLeaf;
-		UDATA arrayletLeafBase = (UDATA)convertPointerFromToken(arrayoidPointer[arrayletIndex]);
+		UDATA arrayletLeafBase = 0;
+		fj9object_t *arrayletLeafSlot = GC_SlotObject::addToSlotAddress(arrayoidPointer, arrayletIndex, compressObjectReferences());
+		if (compressObjectReferences()) {
+			arrayletLeafBase = (UDATA)convertPointerFromToken(*(U_32*)arrayletLeafSlot);
+		} else {
+			arrayletLeafBase = *(UDATA*)arrayletLeafSlot;
+		}
```

Comment:
We have up to 3 'if CR' runtime checks, while it's probably doable with only one :(

Reply:
Something like this?
```
		if (compressObjectReferences()) {
			arrayletLeafBase = (UDATA)convertCompressedPointerFromToken(*((U_32*)arrayoidPointer) + arrayletIndex);
		} else {
			arrayletLeafBase = *(((UDATA*)arrayoidPointer) + arrayletIndex);
		}
```

where convertCompressedPointerFromToken() does just shifting and no 'if CR' check. This helper could be used elsewhere (in readImpl() )